xtokens/src/lib.rs

```diff
@@ -255,13 +312,100 @@ pub mod module {
 
 		/// Ensure has the `dest` has chain part and recipient part.
 		fn ensure_valid_dest(
-			dest: MultiLocation,
+			dest: &MultiLocation,
 		) -> sp_std::result::Result<(MultiLocation, MultiLocation), DispatchError> {
 			if let (Some(dest), Some(recipient)) = (dest.chain_part(), dest.non_chain_part()) {
 				Ok((dest, recipient))
 			} else {
 				Err(Error::<T>::InvalidDest.into())
 			}
 		}
+
+		/// Get the transfer kind.
+		///
+		/// Returns `Err` if `asset` and `dest` combination doesn't make sense,
+		/// else returns a tuple of:
+		/// - `transfer_kind`.
+		/// - asset's `reserve` parachain or relay chain location,
+		/// - `dest` parachain or relay chain location.
+		/// - `recipient` location.
+		fn transfer_kind(
+			asset: &MultiAsset,
+			dest: &MultiLocation,
+		) -> sp_std::result::Result<(TransferKind, MultiLocation, MultiLocation, MultiLocation), DispatchError> {
+			let (dest, recipient) = Self::ensure_valid_dest(dest)?;
+
+			let self_location = T::SelfLocation::get();
+			ensure!(dest != self_location, Error::<T>::NotCrossChainTransfer);
+
+			let reserve = asset.reserve().ok_or(Error::<T>::AssetHasNoReserve)?;
+			let transfer_kind = if reserve == self_location {
+				SelfReserveAsset
+			} else if reserve == dest {
+				ToReserve
+			} else {
+				ToNonReserve
+			};
+			Ok((transfer_kind, dest, reserve, recipient))
+		}
+	}
+
+	// weights
+	impl<T: Config> Pallet<T> {
+		/// Returns weight of `transfer_multiasset` call.
+		fn weight_of_transfer_multiasset(asset: &MultiAsset, dest: &MultiLocation) -> Weight {
+			if let Ok((transfer_kind, dest, _, reserve)) = Self::transfer_kind(asset, dest) {
+				let mut msg = match transfer_kind {
+					SelfReserveAsset => WithdrawAsset {
+						assets: vec![asset.clone()],
+						effects: vec![DepositReserveAsset {
+							assets: vec![All],
+							dest,
+							effects: vec![],
+						}],
+					},
+					ToReserve | ToNonReserve => {
+						WithdrawAsset {
+							assets: vec![asset.clone()],
+							effects: vec![InitiateReserveWithdraw {
+								assets: vec![All],
+								// `dest` is always (equal to) `reserve` in both cases
```

Comment:
That seems wrong?

Reply:
I think in both cases (non self reserve assets) the XCM will be sent to reserve chain to update reserve accounts firstly. If dest is reserve chain, the XCM execution stops there, or the reserve chain will send another XCM to dest notifying the reserve account change.